harvester/harvest.py

```diff
@@ -0,0 +1,497 @@
+        return self._harvest_source
+
+    @property
+    def identifier(self) -> str:
+        return self._identifier
+
+    @property
+    def metadata(self) -> dict:
+        return self._metadata
+
+    @property
+    def metadata_hash(self) -> str:
+        return self._metadata_hash
+
+    @property
+    def operation(self):
+        return self._operation
+
+    @operation.setter
+    def operation(self, value):
+        if value not in ["create", "update", "delete", None]:
+            raise ValueError("Unknown operation being set to record")
+        self._operation = value
+
+    @property
+    def valid(self):
+        return self._valid
+
+    @valid.setter
+    def valid(self, value):
+        if not isinstance(value, bool):
+            raise ValueError("Record validity must be expressed as a boolean")
+        self._valid = value
+
+    @property
+    def validation_msg(self):
+        return self._validation_msg
+
+    @validation_msg.setter
+    def validation_msg(self, value):
+        if not isinstance(value, str):
+            raise ValueError("status must be a string")
+        self._validation_msg = value
+
+    @property
+    def status(self):
+        return self._status
+
+    @status.setter
+    def status(self, value):
+        if not isinstance(value, str):
+            raise ValueError("status must be a string")
+        self._status = value
+
+    # ckanify methods
+    def create_ckan_extra_base(self, *args) -> dict:
+        keys = ["publisher_hierarchy", "resource-type", "publisher"]
+        data = zip(keys, args)
+        return [{"key": d[0], "value": d[1]} for d in data]
+
+    def create_ckan_extras_additions(
+        self, metadata: dict, additions: list[dict]
+    ) -> list[dict]:
+        extras = [
+            "accessLevel",
+            "bureauCode",
+            "identifier",
+            "modified",
+            "programCode",
+            "publisher",
+        ]
+
+        output = []
+
+        for extra in extras:
+            data = {"key": extra, "value": None}
+            val = metadata[extra]
+            if extra == "publisher":
+                data["value"] = val["name"]
+            else:
+                if isinstance(val, list):  # TODO: confirm this is what we want.
+                    val = val[0]
+                data["value"] = val
+            output.append(data)
+
+        return output + additions
+
+    def create_ckan_tags(self, keywords: list[str]) -> list:
+        output = []
+
+        for keyword in keywords:
+            keyword = "-".join(keyword.split())
+            output.append({"name": keyword})
+
+        return output
+
+    def create_ckan_publisher_hierarchy(self, pub_dict: dict, data: list = []):
+        for k, v in pub_dict.items():
+            if k == "name":
+                data.append(v)
+            if isinstance(v, dict):
+                self.create_ckan_publisher_hierarchy(v, data)
+
+        return " > ".join(data[::-1])
```

Comment:
remind me again what `[::-1]` denotes

Reply:
nvm. https://stackoverflow.com/questions/31633635/what-is-the-meaning-of-inta-1-in-python

what exactly is it doing here then?

Reply:
it joins the reversed list of strings together by ` > `. for example
```python
test_publisher_dict = {
                "name": "U.S. Commodity Futures Trading Commission",
                "subOrganizationOf": {
                    "name": "U.S. Government"
                }
            }

# after traversing through all the nested dictionaries "data" will look like this...
data = [ "U.S. Commodity Futures Trading Commission", "U.S. Government" ]

" > ".join(data[::-1]) # produces the string "U.S. Government > U.S. Commodity Futures Trading Commission"


```